File: tests/simulador_pc/validador.py
# ============================================================
# validador.py — Normaliza texto de Whisper → comando válido
#
# Convierte texto transcrito en un comando canónico del juego.
# Incluye filtro de alucinaciones igual que validador.ts del browser.
#
# Usado solo por el simulador cuando necesita validar texto
# (actualmente el browser valida con validador.ts, pero este
# archivo se mantiene como referencia y para posibles futuros usos).
# ============================================================

import unicodedata
import re
from collections import Counter

# Variantes fonéticas por comando
VARIANTES: dict[str, list[str]] = {
    "ROJO":      ["ROJO", "ROJA", "ROXO", "RONJO", "ROCO", "ROSO"],
    "VERDE":     ["VERDE", "BERDE", "BERDI", "VERD", "ERDE", "BIRDE"],
    "AZUL":      ["AZUL", "ASUL", "AZUR", "ASOR", "ASUR"],
    "AMARILLO":  ["AMARILLO", "AMARILLA", "AMARIJO", "MARILLO", "AMARILO", "MARRILLO"],
    "START":     ["START", "EMPIEZA", "INICIA", "COMIENZA", "JUGAR", "ARRANCA",
                  "EMPEZAR", "INICIAR", "COMENZAR", "JUEGA", "ARRANCATE",
                  "EMPIEZE", "INICIE", "INIZI", "INIZIA"],
    "STOP":      ["STOP", "PARA", "PARAR", "TERMINA", "FIN", "SALIR",
                  "TERMINAR", "DETENTE", "ALTO"],
    "PAUSA":     ["PAUSA", "PAUSAR", "ESPERA", "ESPERAR"],
    "REPITE":    ["REPITE", "REPETIR", "REPITA", "REPITELO", "REPITELA"],
    "REINICIAR": ["REINICIAR", "REINICIA", "RESET", "VOLVER",
                  "REINICIATE", "REINICIO"],
    "ARRIBA":    ["ARRIBA", "ARRIVA"],
    "ABAJO":     ["ABAJO", "ABAHO", "AVAJO"],
    "IZQUIERDA": ["IZQUIERDA", "ISKIERDA"],
    "DERECHA":   ["DERECHA", "DERECA"],
    "SI":        ["SI"],
    "NO":        ["NO"],
}

# Frases de dos palabras
FRASES: dict[str, str] = {
    "OTRA VEZ": "REPITE",
    "DE NUEVO": "REPITE",
    "OTRA VES": "REPITE",
}

# Mapa inverso: variante → comando canónico
_MAPA: dict[str, str] = {}
for _cmd, _vars in VARIANTES.items():
    for _v in _vars:
        _MAPA[_v] = _cmd

# Todas las claves del mapa para el filtro de alucinaciones
_VOCAB = set(_MAPA.keys())


def _normalizar(texto: str) -> str:
    """Mayúsculas, sin acentos, sin puntuación."""
    texto = texto.upper()
    texto = unicodedata.normalize("NFD", texto)
    texto = "".join(c for c in texto if unicodedata.category(c) != "Mn")
    texto = re.sub(r"[^A-Z0-9 ]", "", texto)
    return texto.strip()


def _es_alucinacion(texto: str) -> bool:
    """
    Detecta alucinaciones de Whisper:
    - Sin ninguna palabra del vocabulario del juego
    - Repetición de la misma palabra > 3 veces (loop)
    """
    palabras = _normalizar(texto).split()
    if not palabras:
        return True
    # Sin vocabulario del juego → alucinación
    if not (set(palabras) & _VOCAB):
        return True
    # Loop de repetición → alucinación
    conteo = Counter(palabras)
    if conteo.most_common(1)[0][1] > 3:
        return True
    return False


def texto_a_comando(texto: str) -> str:
    """Convierte texto de Whisper en un comando del juego. Retorna 'DESCONOCIDO' si no hay coincidencia."""
    if not texto:
        return "DESCONOCIDO"
    if _es_alucinacion(texto):
        return "DESCONOCIDO"

    norm = _normalizar(texto)
    if not norm:
        return "DESCONOCIDO"

    # 1. Coincidencia exacta
    if norm in _MAPA:
        return _MAPA[norm]

    # 2. Frases de dos palabras
    for frase, cmd in FRASES.items():
        if frase in norm:
            return cmd

    # 3. Palabra por palabra — solo si el texto es corto (≤ 3 palabras)
    #    Evita falsos positivos como "y no hay algo que no hay" → NO
    palabras = norm.split()
    if len(palabras) <= 3:
        for palabra in palabras:
            if palabra in _MAPA:
                return _MAPA[palabra]

    return "DESCONOCIDO"

File: web-panel/lib/validador.ts
// ============================================================
// lib/validador.ts — Normaliza texto de Whisper → comando válido
//
// Convierte el texto transcrito por Whisper WASM en un comando
// canónico del juego (ROJO, START, PAUSA, etc.).
//
// Incluye filtro de alucinaciones: Whisper a veces inventa texto
// cuando no hay voz real. Se descartan frases que no contienen
// ninguna palabra del vocabulario del juego.
// ============================================================

// Variantes fonéticas por comando
const VARIANTES: Record<string, string[]> = {
  ROJO:      ["ROJO", "ROJA", "ROXO", "RONJO", "ROCO", "ROSO"],
  VERDE:     ["VERDE", "BERDE", "BERDI", "VERD", "ERDE", "BIRDE"],
  AZUL:      ["AZUL", "ASUL", "AZUR", "ASOR", "ASUR", "AZUUL"],
  AMARILLO:  ["AMARILLO", "AMARILLA", "AMARIJO", "MARILLO", "AMARILO", "MARRILLO"],
  START:     ["START", "EMPIEZA", "INICIA", "COMIENZA", "JUGAR", "ARRANCA",
              "EMPEZAR", "INICIAR", "COMENZAR", "JUEGA", "ARRANCATE",
              "EMPIEZE", "INICIE", "INIZI", "INIZIA"],
  STOP:      ["STOP", "PARA", "PARAR", "TERMINA", "FIN", "SALIR",
              "TERMINAR", "DETENTE", "ALTO"],
  PAUSA:     ["PAUSA", "PAUSAR", "ESPERA", "ESPERAR"],
  REPITE:    ["REPITE", "REPETIR", "REPITA", "REPITELO", "REPITELA"],
  REINICIAR: ["REINICIAR", "REINICIA", "RESET", "VOLVER", "REINICIATE",
              "REINIZIO", "REINICIO"],
  ARRIBA:    ["ARRIBA", "ARRIVA"],
  ABAJO:     ["ABAJO", "ABAHO", "AVAJO"],
  IZQUIERDA: ["IZQUIERDA", "ISKIERDA"],
  DERECHA:   ["DERECHA", "DERECA"],
  SI:        ["SI", "SÍ", "YES"],
  NO:        ["NO", "NON"],
};

// Frases de dos palabras
const FRASES: Record<string, string> = {
  "OTRA VEZ": "REPITE",
  "DE NUEVO": "REPITE",
  "OTRA VES": "REPITE",
  "VOLVER A": "REPITE",
};

// Mapa inverso: variante → comando canónico
const MAPA_VARIANTES: Record<string, string> = {};
for (const [comando, variantes] of Object.entries(VARIANTES)) {
  for (const v of variantes) {
    MAPA_VARIANTES[v] = comando;
  }
}

// Todas las variantes del vocabulario — para el filtro de alucinaciones
const VOCAB_PALABRAS = new Set(Object.keys(MAPA_VARIANTES));

/**
 * Detecta alucinaciones de Whisper:
 * texto que no contiene ninguna palabra del vocabulario del juego,
 * o que repite la misma palabra más de 3 veces (loop de alucinación).
 */
function esAlucinacion(texto: string): boolean {
  const palabras = normalizar(texto).split(/\s+/).filter(Boolean);
  if (!palabras.length) return true;

  // Sin ninguna palabra del vocabulario → alucinación (ej: "Bienvenidos.", "[Música]")
  if (!palabras.some((p) => VOCAB_PALABRAS.has(p))) return true;

  // Repetición excesiva de la misma palabra → loop de alucinación
  const conteo = new Map<string, number>();
  for (const p of palabras) conteo.set(p, (conteo.get(p) ?? 0) + 1);
  if (Math.max(...conteo.values()) > 3) return true;

  return false;
}

/**
 * Normaliza texto crudo: mayúsculas, sin acentos, sin puntuación.
 */
function normalizar(texto: string): string {
  return texto
    .toUpperCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^A-Z0-9 ]/g, "")
    .trim();
}

/**
 * Convierte texto transcrito por Whisper en un comando del juego.
 * Retorna el comando en mayúsculas o "DESCONOCIDO".
 */
export function textoAComando(texto: string): string {
  if (!texto) return "DESCONOCIDO";
  if (esAlucinacion(texto)) return "DESCONOCIDO";

  const normalizado = normalizar(texto);
  if (!normalizado) return "DESCONOCIDO";

  // 1. Coincidencia exacta del texto completo
  if (MAPA_VARIANTES[normalizado]) return MAPA_VARIANTES[normalizado];

  // 2. Frases de dos palabras
  for (const [frase, comando] of Object.entries(FRASES)) {
    if (normalizado.includes(frase)) return comando;
  }

  // 3. Palabra por palabra — solo si el texto es corto (≤ 3 palabras)
  //    Evita falsos positivos en frases largas como "y no hay algo que no hay" → NO
  const palabras = normalizado.split(/\s+/).filter(Boolean);
  if (palabras.length <= 3) {
    for (const palabra of palabras) {
      if (MAPA_VARIANTES[palabra]) return MAPA_VARIANTES[palabra];
    }
  }

  return "DESCONOCIDO";
}
